chainindex/indexer.go

```diff
@@ -0,0 +1,475 @@
+		return xerrors.Errorf("prepare %s: %w", "insertTipsetMessageStmt", err)
+	}
+	si.tipsetExistsStmt, err = si.db.Prepare(stmtTipsetExists)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "tipsetExistsStmt", err)
+	}
+	si.tipsetUnRevertStmt, err = si.db.Prepare(stmtTipsetUnRevert)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "tipsetUnRevertStmt", err)
+	}
+	si.revertTipsetStmt, err = si.db.Prepare(stmtRevertTipset)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "revertTipsetStmt", err)
+	}
+	si.getMaxNonRevertedTipsetStmt, err = si.db.Prepare(stmtGetMaxNonRevertedTipset)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "getMaxNonRevertedTipsetStmt", err)
+	}
+	si.removeRevertedTipsetsBeforeHeightStmt, err = si.db.Prepare(stmtRemoveRevertedTipsetsBeforeHeight)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "removeRevertedTipsetsBeforeHeightStmt", err)
+	}
+	si.removeTipsetsBeforeHeightStmt, err = si.db.Prepare(stmtRemoveTipsetsBeforeHeight)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "removeTipsetsBeforeHeightStmt", err)
+	}
+	si.deleteEthHashesOlderThanStmt, err = si.db.Prepare(stmtDeleteEthHashesOlderThan)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "deleteEthHashesOlderThanStmt", err)
+	}
+	si.revertTipsetsFromHeightStmt, err = si.db.Prepare(stmtRevertTipsetsFromHeight)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "revertTipsetsFromHeightStmt", err)
+	}
+	si.countMessagesStmt, err = si.db.Prepare(stmtCountMessages)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "countMessagesStmt", err)
+	}
+	si.minNonRevertedHeightStmt, err = si.db.Prepare(stmtMinNonRevertedHeight)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "minNonRevertedHeightStmt", err)
+	}
+	si.tipsetExistsNotRevertedStmt, err = si.db.Prepare(stmtTipsetExistsNotReverted)
+
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "tipsetExistsNotRevertedStmt", err)
+	}
+	si.tipsetExistsNotRevertedStmt, err = si.db.Prepare(stmtTipsetExistsNotReverted)
+	if err != nil {
+		return xerrors.Errorf("prepare %s: %w", "tipsetExistsNotRevertedStmt", err)
+	}
+
+	return nil
+}
+
+func (si *SqliteIndexer) IndexEthTxHash(ctx context.Context, txHash ethtypes.EthHash, msgCid cid.Cid) error {
+	si.closeLk.RLock()
+	if si.closed {
+		return ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	return withTx(ctx, si.db, func(tx *sql.Tx) error {
+		return si.indexEthTxHash(ctx, tx, txHash, msgCid)
+	})
+}
+
+func (si *SqliteIndexer) IndexSignedMessage(ctx context.Context, msg *types.SignedMessage) error {
+	si.closeLk.RLock()
+	if si.closed {
+		return ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	return withTx(ctx, si.db, func(tx *sql.Tx) error {
+		return si.indexSignedMessage(ctx, tx, msg)
+	})
+}
+
+func (si *SqliteIndexer) indexSignedMessage(ctx context.Context, tx *sql.Tx, msg *types.SignedMessage) error {
+	if msg.Signature.Type != crypto.SigTypeDelegated {
```

Comment:
maybe rethink this check and where it should be done? I like that it's so close to execution but a couple of observations:

1. This is after a lock and transaction, you could have returned early from `IndexSignedMessage`.
2. This check is done when you index a tipset before you even get here, but it's _not_ done where you watch the mpool.

I don't mind repeated checks so much, but there's inconsistency and a minor perf hit on the mpool. Perhaps if you just did it once at the top of `IndexSignedMessage` then you get the best of both worlds?